File: lib/haka/lua/lua/parse_ctx.lua
-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at http://mozilla.org/MPL/2.0/.

#include <haka/config.h>


local class = require('class')
local parseResult = require('parse_result')

local parse_ctx_new

#ifdef HAKA_FFI

local ffi = require('ffi')
local ffibinding = require('ffibinding')

ffi.cdef[[
	bool parse_ctx_new_ffi(struct parse_ctx_object *parse_ctx, void *iter);
	void parse_ctx_free(struct parse_ctx *ctx);
	struct lua_ref *parse_ctx_get_ref(void *ctx);

	void parse_ctx_mark(struct parse_ctx *ctx, bool readonly);
	void parse_ctx_unmark(struct parse_ctx *ctx);
	void parse_ctx_pushmark(struct parse_ctx *ctx);
	void parse_ctx_popmark(struct parse_ctx *ctx, bool seek);
	void parse_ctx_seekmark(struct parse_ctx *ctx);
	void parse_ctx_update_error(struct parse_ctx *ctx, const char id[], const char rule[]);
	void parse_ctx_error(struct parse_ctx *ctx, const char desc[]);

	struct parse_ctx {
		int run;
		int node;
		int bitoffset;
	};
]]

ffibinding.create_type{
	cdef = "struct parse_ctx",
	prop = {
	},
	meth = {
		mark = ffi.C.parse_ctx_mark,
		unmark = ffi.C.parse_ctx_unmark,
		pushmark = ffi.C.parse_ctx_pushmark,
		popmark = ffi.C.parse_ctx_popmark,
		seekmark = ffi.C.parse_ctx_seekmark,
		update_error = ffi.C.parse_ctx_update_error,
		error = ffi.C.parse_ctx_error,
	},
	destroy = ffi.C.parse_ctx_free,
	ref = ffi.C.parse_ctx_get_ref,
}

parse_ctx_new = ffibinding.object_wrapper("struct parse_ctx", ffibinding.handle_error(ffi.C.parse_ctx_new_ffi), true)

#else

parse_ctx_new = haka.parse_ctx
haka.parse_ctx = nil

#endif

--
-- Parse C Context
--

CContext = class.class('ParseCContext')

CContext.property.current_init = {
	get = function (self)
		if self._initresults then
			return self._initresults[#self._initresults]
		end
	end
}

CContext.property.init = {
	get = function (self) return self._initresults ~= nil end
}

CContext.property.retain_mark = {
	get = function (self)
		return self._retain_mark[#self._retain_mark]
	end
}

CContext.property._bitoffset = {
	get = function (self)
		return self._ctx.bitoffset
	end,
	set = function (self, val)
		self._ctx.bitoffset = val
	end
}

local function revalidate(self)
	local validate = self._validate
	self._validate = {}
	for f, arg in pairs(validate) do
		f(arg)
	end
end

function CContext.method:__init(iter, init)
	self._ctx = parse_ctx_new(iter)
	self.iter = iter
	self._catches = {}
	self._results = {}
	self._validate = {}

	if init then
		self._initresults = { init }
	end

	self.iter.meter = 0
end

function CContext.method:mark(readonly)
	return self._ctx:mark(readonly)
end

function CContext.method:unmark()
	return self._ctx:unmark()
end

function CContext.method:pushmark()
	return self._ctx:pushmark()
end

function CContext.method:popmark(seek)
	local seek = seek or false
	return self._ctx:popmark(seek)
end

function CContext.method:seekmark()
	return self._ctx:seekmark()
end

function CContext.method:error(desc, ...)
	local desc = string.format(desc, ...)
	return self._ctx:error(desc)
end

function CContext.method:update_error(id, rule)
	local id = id or "<unknown>"
	local rule = rule or "<unknown>"
	return self._ctx:update_error(id, rule)
end

function CContext.method:result(idx)
	idx = idx or -1

	if idx < 0 then
		idx = #self._results + 1 + idx
		if idx < 0 then
			error("invalid result index")
		end
	else
		if idx > #self._results then
			error("invalid result index")
		end
	end

	if idx <= 0 then
		error("invalid result index")
	end

	return self._results[idx]
end

function CContext.method:update(iter)
	self.iter:move_to(iter)
end

function CContext.method:lookahead()
	local iter = self.iter:copy()
	local sub = self.iter:sub(1)
	if sub then
		local la = sub:asnumber()
		self:update(iter)
		return la
	else
		return -1
	end
end

function CContext.method:init(entity, all)
	if entity.resultclass then
		self:push(entity.resultclass:new())
	else
		self:push(parseResult.Result:new())
	end
	self:result(1).validate = revalidate

	if all then self._level = 1
	else self._level = 0 end

	self._level_exit = 0
	self._error = nil
end

function CContext.method:pop()
	assert(#self._results > 0)

	self._results[#self._results] = nil

	if self._initresults then
		self._initresults[#self._initresults] = nil
	end
end

function CContext.method:push(result, name)
	local new = result or parseResult.Result:new()
	rawset(new, '_validate', self._validate)
	self._results[#self._results+1] = new
	if self._initresults then
		local curinit = self._initresults[#self._initresults]
		if curinit then
			self._initresults[#self._initresults+1] = curinit[name]
		else
			self._initresults[#self._initresults+1] = nil
		end
	end
	return new
end

return CContext

File: lib/haka/lua/parse_ctx.c
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#include <assert.h>

#include <haka/config.h>
#include <haka/lua/parse_ctx.h>
#include <haka/error.h>

#ifdef HAKA_DEBUG
#define POOL_MEM_GUARD(pool) do {                                              \
	memset(&(pool)->el[(pool)->count], 0,                                  \
			((pool)->alloc - (pool)->count) * (pool)->size);       \
} while(0)
#else
#define POOL_MEM_GUARD(pool) do { } while(0)
#endif

#define POOL_INIT(elem, pool, init) do {                                       \
	(pool)->alloc = init;                                                  \
	(pool)->size = sizeof(elem);                                           \
	(pool)->el = calloc((pool)->alloc, (pool)->size);                      \
	if (!(pool)->el) {                                                     \
		error("memory error");                                         \
		return;                                                        \
	}                                                                      \
	(pool)->count = 0;                                                     \
	POOL_MEM_GUARD(pool);                                                  \
} while(0)

#define POOL_FREE(pool) do {                                                   \
	(pool)->alloc = 0;                                                     \
	free((pool)->el);                                                      \
	(pool)->count = 0;                                                     \
} while(0)

#define POOL_REALLOC(pool) do {                                                \
	if((pool)->count >= (pool)->alloc) {                                   \
		(pool)->alloc *= 2;                                            \
		(pool)->el = realloc((pool)->el, (pool)->alloc * (pool)->size);\
		if (!(pool)->el) {                                             \
			error("memory error");                                 \
			return;                                                \
		}                                                              \
		POOL_MEM_GUARD(pool);                                          \
	}                                                                      \
} while(0)

struct parse_ctx *parse_ctx_new(struct vbuffer_iterator *iter)
{
	struct parse_ctx *ctx = malloc(sizeof(struct parse_ctx));
	if (!ctx) {
		error("memory error");
		return NULL;
	}

	parse_ctx_init(ctx, iter);

	return ctx;
}

void parse_ctx_init(struct parse_ctx *ctx, struct vbuffer_iterator *iter)
{
	assert(ctx);

	ctx->run                 = true;
	ctx->current             = 1;
	ctx->next                = 1;
	ctx->lua_object          = lua_object_init;
	ctx->iter                = iter;
	ctx->compound_level      = 0;
	ctx->recurs_finish_level = 0;
	ctx->recurs_count        = 0;
	ctx->bitoffset           = 0;
	ctx->error.isset         = false;
	ctx->error.desc          = NULL;
	ctx->error.node          = 0;

	memset(ctx->recurs, 0, RECURS_MAX*sizeof(struct recurs));

	POOL_INIT(struct mark, &ctx->marks, INIT_MARKS_SIZE);
	POOL_INIT(struct catch, &ctx->catches, INIT_CATCHES_SIZE);
	POOL_INIT(struct vbuffer_iterator, &ctx->retains, INIT_RETAINS_SIZE);
}

void parse_ctx_free(struct parse_ctx *ctx)
{
	POOL_FREE(&ctx->marks);
	POOL_FREE(&ctx->catches);
	POOL_FREE(&ctx->retains);
	free(ctx->error.desc);
	free(ctx);
}

void parse_ctx_mark(struct parse_ctx *ctx, bool readonly)
{
	vbuffer_iterator_copy(ctx->iter, &ctx->retains.el[ctx->retains.count]);
	vbuffer_iterator_mark(&ctx->retains.el[ctx->retains.count], readonly);
	ctx->retains.count++;
	POOL_REALLOC(&ctx->retains);
}

void parse_ctx_unmark(struct parse_ctx *ctx)
{
	ctx->retains.count--;
	vbuffer_iterator_unmark(&ctx->retains.el[ctx->retains.count]);

#ifdef HAKA_DEBUG
	memset(&ctx->retains.el[ctx->retains.count], 0, sizeof(struct vbuffer_iterator));
#endif /* HAKA_DEBUG */
}

bool parse_ctx_get_mark(struct parse_ctx *ctx, struct vbuffer_iterator *iter)
{
	if (ctx->retains.count == 0) return false;

	vbuffer_iterator_copy(&ctx->retains.el[ctx->retains.count-1], iter);
	return true;
}

void parse_ctx_pushmark(struct parse_ctx *ctx)
{
	struct mark *mark = &ctx->marks.el[ctx->marks.count];

	vbuffer_iterator_copy(ctx->iter, &mark->iter);
	mark->bitoffset = ctx->bitoffset;
	mark->max_meter = 0;
	vbuffer_iterator_copy(ctx->iter, &mark->max_iter);
	mark->max_bitoffset = ctx->bitoffset;

	ctx->marks.count++;
	POOL_REALLOC(&ctx->marks);
}

void parse_ctx_popmark(struct parse_ctx *ctx, bool seek)
{
	ctx->marks.count--;

	if (seek) {
		struct mark *mark = &ctx->marks.el[ctx->marks.count];
		vbuffer_iterator_move(ctx->iter, &mark->max_iter);
		ctx->bitoffset = mark->max_bitoffset;
	}

#ifdef HAKA_DEBUG
	memset(&ctx->marks.el[ctx->marks.count], 0, sizeof(struct mark));
#endif /* HAKA_DEBUG */
}

void parse_ctx_seekmark(struct parse_ctx *ctx)
{
	struct mark *mark = &ctx->marks.el[ctx->marks.count-1];
	int len = ctx->iter->meter - mark->iter.meter;

	if (len >= mark->max_meter) {
		vbuffer_iterator_clear(ctx->iter);
		vbuffer_iterator_copy(&mark->max_iter, ctx->iter);
		vbuffer_iterator_register(ctx->iter);
		mark->max_meter = len;
		if (ctx->bitoffset > mark->max_bitoffset) {
			mark->max_bitoffset = ctx->bitoffset;
		}
	}

	vbuffer_iterator_move(ctx->iter, &mark->max_iter);
	ctx->bitoffset = mark->bitoffset;
}

void parse_ctx_pushcatch(struct parse_ctx *ctx, int node)
{
	struct catch *catch = &ctx->catches.el[ctx->catches.count];

	/* Should have created a ctx result */
	parse_ctx_mark(ctx, false);
	parse_ctx_pushmark(ctx);

	catch->node = node;
	catch->retain_count = ctx->retains.count;
	catch->mark_count = ctx->marks.count;

	ctx->catches.count++;
	POOL_REALLOC(&ctx->catches);
}

bool parse_ctx_catch(struct parse_ctx *ctx)
{
	struct catch *catch;
	int i;

	if (ctx->catches.count == 0) {
		ctx->next = FINISH;
		return false;
	}

	catch = &ctx->catches.el[--ctx->catches.count];

	/* Unmark each retain started in failing case */
	for(i = ctx->retains.count; i > catch->retain_count; i--) {
		parse_ctx_unmark(ctx);
	}

	/* remove all marks done in failing case */
	for (i = ctx->marks.count; i > catch->mark_count; i--) {
		parse_ctx_popmark(ctx, false);
	}


	assert(!"Not Yet Implemented");
	/* remove all result ctx */
	/* for (i = ctx->result_count; i > catch->result_count; i--) {
		// TODO parse_ctx_pop();
	} */

	/* Also remove result created by Try */
	/* Should be done in Try entity but we can't */
	// parse_ctx_pop();

	parse_ctx_seekmark(ctx);
	parse_ctx_popcatch(ctx);

	ctx->next = catch->node;

	return true;
}

void parse_ctx_popcatch(struct parse_ctx *ctx)
{
	ctx->catches.count--;

	parse_ctx_popmark(ctx, false);
	parse_ctx_unmark(ctx);

#ifdef HAKA_DEBUG
	memset(&ctx->catches.el[ctx->catches.count], 0, sizeof(struct catch));
#endif /* HAKA_DEBUG */

}

void parse_ctx_error(struct parse_ctx *ctx, const char desc[])
{
	if (ctx->error.isset) {
		error("multiple parse errors raised");
		return;
	}

	ctx->error.node = ctx->current;
	ctx->error.isset = true;
	vbuffer_iterator_copy(ctx->iter, &ctx->error.iter);
	ctx->error.desc = strdup(desc);
}

bool parse_ctx_haserror(struct parse_ctx *ctx)
{
	return ctx->error.isset;
}

#ifdef HAKA_FFI

extern void *lua_get_swigdata(void *ptr);

bool parse_ctx_new_ffi(struct ffi_object *parse_ctx, void *_iter)
{
	struct parse_ctx *ctx = parse_ctx_new((struct vbuffer_iterator *)lua_get_swigdata(_iter));
	if (!ctx) {
		return false;
	}
	else {
		parse_ctx->ref = &ctx->lua_object.ref;
		parse_ctx->ptr = ctx;
		return true;
	}
}

bool parse_ctx_get_mark_ffi(struct parse_ctx *ctx, void *_iter)
{
	return parse_ctx_get_mark(ctx, (struct vbuffer_iterator *)lua_get_swigdata(_iter));
}

struct lua_ref *parse_ctx_get_ref(void *_ctx)
{
	struct parse_ctx *ctx = (struct parse_ctx *)_ctx;
	return &ctx->lua_object.ref;
}
#endif /* HAKA_FFI */
